### ISSUE TITLE
Steinhafels - Correct City/State errors on customers and orders at order save time.

### ISSUE THREAD

**Issue by kentbower** [kbower/tickettest1#2375]
_Imported from trac ticket 2375.
Created by **dzuern**
Opened in trac: Jun 13, 2013, 2:23:39 PM
Last modified in trac: Jul 24, 2013, 11:17:03 AM_
***
_by dzuern:_
Because of issues where users mis-key the zip code and/or the city name, we need to enforce correct entry while allowing flexibility for non configured zips.  
We can take care of this with a warning that will fix a mismatch and warn them at save time. We always try to pull in the correct city for the zip code entered even before the save. If you went back in as Denise did and try to make it wrong, we can fix it and add the warning that the city does not match the configured Zip at save time. We can also allow it to save (customer) with any city if the zip is not in the system, but if this is a Delivery, the ship to will have to be valid to allow the save. So I believe all we need to add new here is the logic to fix and warn at save time that the city/state on either the customer or ship to record does not match the entered valid system zip code. The message can say "City/State updated to match the entered zip code".

***
* criley attached [ticket2375.patch](https://trac.retailarchitects.com/trac/attachment/ticket/2375/ticket2375.patch) on Jul 12, 2013, 11:24:58 AM

**Comment by kentbower** [comment]
_**sjohnson** trac comment on Jun 17, 2013, 2:10:24 PM:_

Denise wrote:
This sounds good. Just to confirm here is what needs to happen:
    Verify that City field is a match for the Zip (so slsp cannot change the City Name, like I did in testing).
    Allow "any" zip/city to be entered IF a pickup.
    If order is a delivery the zip/city MUST be valid in our delivery area.
If everyone agree, please make the change.

**Comment by kentbower** [comment]
_**kbower** trac comment on Jul 12, 2013, 11:43:52 AM:_

This doesn't address _If order is a delivery the zip/city MUST be valid in our delivery area_, but I assume the delivery calendar is already taking care of that, so I don't know why Denise wrote that...

I don't think I agree with the check: `isinstance(order, _ReservableOrder)` because a Return or ServiceOrder should also update this correctly if it is a Delivery.

As written, the rule will attempt to change already closed orders and cause an Exception.  So instead of `isinstance(order, _ReservableOrder)`, please use `order._original_status_ == u'O'`.

Please check `order.deliverytype != u'P'` _before_ you check `order.shipcity != order.shipzipcodeobj.city` to potentially save a round trip to the database for pickups (there is no need to check city if == u'P').

Looks very good!

**Comment by kentbower** [comment]
_**kbower** trac comment on Jul 12, 2013, 11:51:57 AM:_

Also, "city" is nullable.  We certainly don't want to update it to NULL if the zipcode record's city is NULL.

So, change both rules to:
```diff
Index: src/appserver/pylotengine/appsetup/customerdata/stn.py
===================================================================
--- src/appserver/pylotengine/appsetup/customerdata/stn.py (revision 4335)
+++ src/appserver/pylotengine/appsetup/customerdata/stn.py (working copy)
@@ -139,7 +139,12 @@
 capitalize('shipcity')
 ''',),
 
-
+        ('Order','validate_zip_city','4',"Ensure that the order's ZIP code belongs to the given city.",'Y',
+'''
+if order._original_status_ == u'O' and \\
+   order.deliverytype != u'P' and \\
+   order.shipzipcodeobj.city and \\
+   order.shipcity != order.shipzipcodeobj.city:
+    order.shipcity = order.shipzipcodeobj.city
+    order.warning('The given city was updated to match the given ZIP code', 'shipcity')
+''',),
 ##############
 # OrderDetail
 
@@ -181,6 +186,13 @@
 Rule.system_rule('format_phonenumber',obj=customer,fieldname='businessphone')
 ''',),
 
+        ('Customer','validate_zip_city','4',"Ensure that the customer's ZIP code belongs to the given city.",'Y',
+'''
+if customer.custzipcode.city and customer.city != customer.custzipcode.city:
+    customer.city = customer.custzipcode.city
+    customer.warning('The given city was updated to match the given ZIP code', 'city')
+''',),
+
 ################
 # Order Rules
```

**Comment by kentbower** [comment]
_**kbower** trac comment on Jul 12, 2013, 11:56:41 AM:_

Also:
 * orders.shipzipcode is nullable
 * customers.zipcode is nullable

The rules will crash if these are null when they run, so in the if for Order rule you need to add `and order.shipzipcodeobj ...` and in the Customer rule you need to add `if customer.custzipcode and ...`

**Comment by kentbower** [comment]
_**kbower** trac comment on Jul 12, 2013, 11:59:57 AM:_

Also:
The ticket mentions city/**state**... you've only implemented city.  If the city is wrong, the state might be also, I suppose, so you'll need to include that in your changes.

**Comment by kentbower** [comment]
_**criley** trac comment on Jul 24, 2013, 11:17:03 AM:_

(Changeset [[4371]](../commit/5afadbe32f489b29e8ec5d9155106f5c4da2e74e))
* Fixes #2375 (Correct City/State errors on customers and orders at order save time) #stn

**Comment by kentbower** [comment]
_**criley** trac status change on Jul 24, 2013, 11:17:03 AM:_ **closed**
